### PR TITLE
Add !lgc.xfb.state metadata

### PR DESCRIPTION
**Pull request by xuechen417**
We're now preserving XFB state in explicit metadata so that we can
handle the case where there is an XFB output that is never written to.
Meanwhile, the the redudant XFB-related information of
`lgc::ResourceUsage::{enableXfb,xfbStrides,streamXfbBuffers}` is
removed.
The detailed dicussions are in issue(https://github.com/GPUOpen-Drivers/llpc/issues/2109).

See:dEQP-VK.transform_feedback.simple.multiquery_omit_write*
      dEQP-VK.transform_feedback.simple.query_omit_write_0*

### CODE REVIEW

**Comment by xuechen417** [comment]
This PR is splitted from a counterpart PR in the internal repo.

**Comment by amdvlk-admin** [comment]
### Test summary for commit b373085a3fa0d76733298716997d7f763443e536
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/InOutBuilder.cpp [hunk 1]
```diff
@@ -739,9 +739,10 @@ Instruction *InOutBuilder::CreateWriteXfbOutput(Value *valueToWrite, bool isBuil
   unsigned streamId = outputInfo.hasStreamId() ? outputInfo.getStreamId() : 0;
   assert(xfbBuffer < MaxTransformFeedbackBuffers);
   assert(streamId < MaxGsStreams);
-  resUsage->inOutUsage.xfbStrides[xfbBuffer] = xfbStride;
-  resUsage->inOutUsage.enableXfb = true;
-  resUsage->inOutUsage.streamXfbBuffers[streamId] |= 1 << xfbBuffer;
+  auto &xfbStrides = getPipelineState()->getXfbBufferStrides();
+  auto &streamXfbBuffers = getPipelineState()->getStreamXfbBuffers();
+  xfbStrides[xfbBuffer] = xfbStride;
+  streamXfbBuffers[streamId] |= 1 << xfbBuffer;
```
Comment (nhaehnle) [inline lgc/builder/InOutBuilder.cpp:745]: This shouldn't be needed anymore.

**Review by nhaehnle — COMMENTED** [review]
Thank you, this looks pretty good. I do have some detail comments inside.

Also, the commit message doesn't make sense for this change. I know it did for the whole change that's still internal, but here we're not changing anything about XFB query counts. So the commit title should be something like "Add !lgc.xfb.state metadata" with a comment that we're now preserving XFB state in explicit metadata so that we can handle the case where there is an XFB output that is never written to.

Referencing the relevant CTS cases still makes sense, though I'd call it "See <test cases>" instead of "Fixes", because they aren't broken yet as far as the gpuopen dev branch is concerned :)

File: lgc/builder/InOutBuilder.cpp (hunk 1 above)
Comment (xuechen417) [reply lgc/builder/InOutBuilder.cpp:745]: Without updating `xfbstrides` and `streamXfbBuffers`, the block array cases will crash. 
Regression on _dEQP-VK.transform_feedback.fuzz.2_level_struct_array.instance_array_vertex_ [blockArray.zip](https://github.com/GPUOpen-Drivers/llpc/files/10051905/blockArray.zip)

```
#version 450
struct S
{
	highp uvec3 a; // unassigned
	mediump mat2 b;
};
layout(location = 0, xfb_buffer = 0, xfb_offset = 0, xfb_stride = 124) out Block // sizeInBytes=124 sizeInLocations=14
{
	mediump int u;
	S s[2][2];
	mediump vec2 v;
} block[2];
void main (void)
{
	block[0].u = -89;
	block[0].s[0][0].b = mat2(30.0, -70.0, -60.0, 72.0);
	block[0].s[0][1].b = mat2(-102.0, -100.0, 84.0, -31.0);
	block[0].s[1][0].b = mat2(68.0, 104.0, 14.0, 83.0);
	block[0].s[1][1].b = mat2(-117.0, 75.0, 12.0, -18.0);
	block[0].v = vec2(-101.0f, 100.0f);
	block[1].u = -34;
	block[1].s[0][0].b = mat2(-35.0, -43.0, 48.0, 4.0);
	block[1].s[0][1].b = mat2(-60.0, 37.0, 52.0, 98.0);
	block[1].s[1][0].b = mat2(-80.0, -62.0, -54.0, -79.0);
	block[1].s[1][1].b = mat2(3.0, -97.0, -118.0, -108.0);
	block[1].v = vec2(-120.0f, 48.0f);
}
```
In SPIRVReader, we can only get `XfbBuffer 0` used. After `lowerOutput`, `block[0]` is streamed out `XbfBuffer 0` and `block[1]` is streamed out `XfbBuffer 1`. See the effect of parameter `xfbBufferAdjust` of `addCallInstForOutputExport`.

File: lgc/patch/PatchEntryPointMutate.cpp [hunk 2]
```diff
@@ -1233,7 +1233,7 @@ void PatchEntryPointMutate::addSpecialUserDataArgs(SmallVectorImpl<UserDataArg>
   // Allocate register for stream-out buffer table, to go before the user data node args (unlike all the ones
   // above, which go after the user data node args).
   if (userDataUsage->usesStreamOutTable || userDataUsage->isSpecialUserDataUsed(UserDataMapping::StreamOutTable)) {
-    if (enableNgg || !m_pipelineState->getShaderResourceUsage(ShaderStageCopyShader)->inOutUsage.enableXfb) {
+    if (enableNgg || (!m_pipelineState->hasShaderStage(ShaderStageCopyShader) && m_pipelineState->enableXfb())) {
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply lgc/patch/PatchEntryPointMutate.cpp:1236]: Fixed

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -7087,6 +7090,44 @@ bool SPIRVToLLVM::translate(ExecutionModel entryExecModel, const char *entryName
...
+  // StreamId in EmitVertex)
+  if (enableXfb) {
+    const unsigned bufferCount = 8;
+    int xfbState[bufferCount] = {-1, 0, -1, 0, -1, 0, -1, 0};
+    for (unsigned i = 0, e = m_bm->getNumVariables(); i != e; ++i) {
+      auto bv = m_bm->getVariable(i);
+      if (bv->getStorageClass() == StorageClassOutput) {
+        SPIRVWord xfbBuffer = SPIRVID_INVALID;
+        if (bv->hasDecorate(DecorationXfbBuffer, 0, &xfbBuffer)) {
+          const unsigned indexOfStream = 2 * xfbBuffer;
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:7107]: Fixed.

**Review by xuechen417 — COMMENTED** [review]
V1: Update commit message and comment.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 382329ca791a1a459b98a758500288910c3f839b
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/InOutBuilder.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/builder/InOutBuilder.cpp:745]: Interesting, thank you, I hadn't realized that.

Though now I'm curious: what is the desired behavior if `block[1]` is never written in the shader? Shouldn't the write offset of the corresponding xfbBuffer still be incremented, and doesn't that also require capturing the information in SPIRVReader?

**Comment by qiaojbao** [comment]
retest this please

**Comment by qiaojbao** [comment]
### Test summary for commit 382329ca791a1a459b98a758500288910c3f839b
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
At least for me there is some bug in GitHub's UI that hides the latest comment I made about XfB block output arrays. I'm copying a slightly edited version of it here:

> Interesting, thank you, I hadn't realized that.
>
> Though now I'm curious: what is the desired behavior if block[1] is never written in the shader? Shouldn't the write offset of the corresponding xfbBuffer still be incremented, and doesn't that also require capturing the information about the block array size in SPIRVReader?

**Comment by xuechen417** [comment]
> At least for me there is some bug in GitHub's UI that hides the latest comment I made about XfB block output arrays. I'm copying a slightly edited version of it here:
> 
> > Interesting, thank you, I hadn't realized that.
> > Though now I'm curious: what is the desired behavior if block[1] is never written in the shader? Shouldn't the write offset of the corresponding xfbBuffer still be incremented, and doesn't that also require capturing the information about the block array size in SPIRVReader?

Good questions. I wil take some tests and discuss with you later.

**Comment by nhaehnle** [comment]
Hmm, I see now what's happening. In SpirvLowerGlobal, we create proxy allocas for the output variables and rewrite the stores to store into the proxy. We then create `write.xfb.output` operations for the entire proxy. So, even if the original shader did not write to `block[1]`, the generated stores of the proxy will.

Still, I'd prefer for us to be consistent and set up the metadata entirely in the SPIRVReader.

**Comment by xuechen417** [comment]
> Hmm, I see now what's happening. In SpirvLowerGlobal, we create proxy allocas for the output variables and rewrite the stores to store into the proxy. We then create `write.xfb.output` operations for the entire proxy. So, even if the original shader did not write to `block[1]`, the generated stores of the proxy will.
> 
> Still, I'd prefer for us to be consistent and set up the metadata entirely in the SPIRVReader.

Thank you for the explanation of the root cause. I will try to handle it in SPIRVReader.

**Comment by xuechen417** [comment]
V2: Set up !lgc.xfb.state metadata entirely in the SPIRVReader.

**Review by nhaehnle — DISMISSED** [review]
Thank you, LGTM!

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7dbd3ba4e146e0f6a5605a15f04852669bf5e177
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by xuechen417** [comment]
V3: Resolve confilts

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1fb8b20e4dc6add8158975fe88a2fd23afe74257
<details><summary>CTS tests (Failed: 0/207253)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70466 (55.2%)</li>
<li>Failed: 0/70466 (0.0%)</li>
<li>Not Supported: 31561/70466 (44.8%)</li><li>Warnings: 0/70466 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70474 (50.7%)</li>
<li>Failed: 0/70474 (0.0%)</li>
<li>Not Supported: 34724/70474 (49.3%)</li><li>Warnings: 0/70474 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66313 (50.5%)</li>
<li>Failed: 0/66313 (0.0%)</li>
<li>Not Supported: 32837/66313 (49.5%)</li><li>Warnings: 0/66313 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
Can we have an llpc lit test (starting with simplified spir-v that does the not-writing-to-xfb-output thing) please?

**Comment by xuechen417** [comment]
> Can we have an llpc lit test (starting with simplified spir-v that does the not-writing-to-xfb-output thing) please?

Sure, I will add one.

**Comment by xuechen417** [comment]
V4: Add a lit test to check `!lgc.xfb.state`. @nhaehnle It seems the script update_llpc_test_checks.py cannot check the metadata. So I have to manually write the check pattens.

**Comment by amdvlk-admin** [comment]
### Test summary for commit acd647f8864a0db68c8b1c1127e831c71ef44de6
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by xuechen417** [comment]
> > @nhaehnle It seems the script update_llpc_test_checks.py cannot check the metadata. So I have to manually write the check pattens.
> 
> Have you tried with the `--check-globals` option?

Thanks! I missed that option. It works well. The test is updated by the script now.

**Comment by xuechen417** [comment]
V5: Use script to generate FileCheck statement for the test `TestXfbStateMetadata.geom`

**Comment by xuechen417** [comment]
V6: Fix failure of TestXfbStateMetadata.geom

**Review by nhaehnle — APPROVED** [review]
We should improve the script to handle these cases better, but for now I think this test is okay.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f825dbbb2f237413343a811d01018d9d009b2444
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**